src/Components/Web/src/DynamicRootComponentConfiguration.cs

```diff
@@ -0,0 +1,17 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace Microsoft.AspNetCore.Components.Web
+{
+    /// <summary>
+    /// Configures options for allowing JavaScript to add root components dynamically.
+    /// </summary>
+    public abstract class DynamicRootComponentConfiguration
+    {
+        /// <summary>
+        /// Marks the specified component type as allowed for instantiation from JavaScript.
+        /// </summary>
+        /// <typeparam name="TComponent"></typeparam>
+        public abstract void Register<TComponent>(string name) where TComponent : IComponent;
+    }
```

Comment:
The point of this abstract class is to have something that a component library could build an extension method on to register its dynamic components, and hence be agnostic to the hosting model.

The concrete subclass is shared-source into each of the hosting models so we can avoid exposing some unwanted public APIs across layers.